docs/GDPs/gdp-0002.rst

```diff
@@ -0,0 +1,342 @@
+
+We are currently working around the lack of a first-class feature by using the `domain` and `origin` keyword arguments to stencil calls, but doing so leads to inefficient code that is difficult for model developers to read.
+To illustrate the need for such a feature, consider a snippet of the model that computes a variable `ub` different based on location in the grid:
+
+.. code-block:: python
+
+    FIELD_TYPE=gtscript.Field[np.float]​
+    @gtscript.stencil(backend=backend_type)​
+    def main_ub(uc: FIELD_TYPE, vc: FIELD_TYPE, cosa: FIELD_TYPE, rsina: FIELD_TYPE, ub: FIELD_TYPE, dt5: float):​
+        with computation(PARALLEL), interval(...):​
+            ub = dt5 * (uc[0, -1, 0] + uc - ​(vc[-1, 0, 0] + vc) * cosa) * rsina​
+
+    @gtscript.stencil(backend=backend_type)​
+    def x_edge_ub(ut: FIELD_TYPE, ub: FIELD_TYPE, dt5: float)::
+        with computation(PARALLEL), interval(...):​
+            ub = dt5 * (ut[0, -1, 0] + ut)​
+
+    @gtscript.stencil(backend=backend_type)​
+    def y_edge_ub(ut: FIELD_TYPE, ub: FIELD_TYPE, *, dt4: float):
+        with computation(PARALLEL), interval(...):
+            ub = dt4 * (-ut[0, -2, 0] + 3.0 * (ut[0, -1, 0] + ut) - ut[0, 1, 0])
+
+    # Requires new python function to call stencils on each region
+    def compute_ub(uc, vc, ut, ub, dt5, dt4, grid):​
+        domain_y_edge = (grid.ni, 1, grid.nz)
+        domain_x_edge = (1, grid.nj, grid.nz)
+        main_ub(uc, vc, grid.cosa, grid.rsina, ub, dt5=dt5, ​
+                origin=(grid.local_istart, grid.local_jstart, 0),
+                domain=(grid.ni, grid.nj, grid.nz))​
+        if grid.west_edge:​
+            x_edge_ub(ut, ub, dt5=dt5, ​origin=(grid.local_istart, grid.local_jstart, 0), ​domain=domain_x_edge)​
+        if grid.south_edge:
+            y_edge(ut, ub, dt4=dt4, origin=(grid.local_istart, grid.local_jstart, 0), domain=domain_y_edge)
+        if grid.north_edge:
+            y_edge_ub(ut, ub, dt4=dt4, origin=(grid.local_istart, grid.local_jend, 0), domain=domain_y_edge)
+        if grid.east_edge:
+            x_edge_ub(ut, ub, dt5=dt5, origin=(grid.local_iend, grid.local_jstart, 0), domain=domain_x_edge)
+
+This example motivates a number of requirements for the feature. It must be able to
+
+1. specify regions in the context of a global iteration space (for global boundary conditions) using different `with` blocks inside a single stencil
+2. specify arbitrary sub-regions using a clean abstraction
+3. specify cleanly regions at and near the edge of compute domains
+4. read from and write to fields inside halo regions
+
+In the `detailed description` section below we offer a few alternatives, but all these involve adding a `region()` specifier that allows for specifying subsets of the horizontal iteration space.
+Using this, the example above is transformed into:
+
+.. code-block:: python
+
+    @gtscript.stencil(backend=backend_type, partitioner=part_type) # Single stencil
```

Comment:
I believe there is no need to pass `partitioner` as an explicit argument here. Partitioner is linked to this code through the user definitions of WestEdge etc.

Reply:
Leftover code. Thanks for catching!